internal/configuration/settings.go

```diff
@@ -284,11 +284,11 @@ func (s *Settings) handleUpdateEvent(_ interface{}, newValue interface{}) {
 
 	tlsMode, found := configMap.Data["tls-mode"]
 	if found {
-		s.TlsMode = tlsMode
+		s.TlsMode = TLSModeMap[tlsMode]
```

Comment:
If the value provided is not defined in `TLSModeMap`, we're going to get `0` here which **should** correspond to `NoTLS` enum but creates a funny invisible dependency on the order in which those constants are defined in the `configuration` package.

I think it would be better to explicitly define what comes back (for now - this should be the error case but as we have discussed [here](https://github.com/nginxinc/nginx-loadbalancer-kubernetes/issues/143#issuecomment-1847434139)).

So something like 
```go
rawTLSMode, found := configMap.Data["tls-mode"]
if found && tlsMode, enumExists := TLSModeMap[rawTLSMode]; enumExists {
  s.TlsMode = TLSModeMap[tlsMode]
}
```

Or some variation of that (the above code has the weakness of making the `else` case cover two different error modes).

Reply:
Yeah, the "default to the iota value enum" threw me at first. I monkeyed around with using another value for iota (the first enum) to try and capture the error condition, but it was super clunky and not intuitive.

I see what you're getting at. Yep.

Another point (I'll move all of this to another issue tomorrow); when you mentioned not applying a configuration if the TLS Mode is invalid, did you mean the WHOLE configuration, or just the TLS Mode configuration?

I am not sure disregarding the whole configuration is a good idea...

Reply:
AHA. Killing two birds with one stone:

```
	tlsMode, err := validateTlsMode(configMap)
	if err != nil {
		logrus.Errorf("There was an error with the configured TLS Mode, no changes applied: %s", err)
	} else {
		s.TlsMode = tlsMode
	}
```

with the `validateTlsMode` function as follows:

```
func validateTlsMode(configMap *corev1.ConfigMap) (TLSMode, error) {
	tlsConfigMode, tlsConfigModeFound := configMap.Data["tls-mode"]
	if !tlsConfigModeFound {
		return NoTLS, fmt.Errorf(`tls-mode key not found in ConfigMap`)
	}

	if tlsMode, tlsModeFound := TLSModeMap[tlsConfigMode]; tlsModeFound {
		return tlsMode, nil
	}

	return NoTLS, fmt.Errorf(`invalid tls-mode value: %s`, tlsConfigMode)
}

```

We just need to figure out how to ensure the user is slapped upside the head when that Error fires. The simplest solution is to document it, suggest a simple way to monitor this, and trust the user will follow directions.

Reply:
When I looked at this more closely I realized what you were saying about handling two error conditions. That led me to extract a method that could distinguish between them and report back accordingly. I think this more clearly establishes the possible failure modes. 


Incidentally, the conditional in the original comment :

```
rawTLSMode, found := configMap.Data["tls-mode"]
if found && tlsMode, enumExists := TLSModeMap[rawTLSMode]; enumExists {
  s.TlsMode = TLSModeMap[tlsMode]
}
```

Would need to be structured thusly:

```
rawTLSMode, found := configMap.Data["tls-mode"]
if tlsMode, enumExists := TLSModeMap[rawTLSMode]; found && enumExists {
  s.TlsMode = TLSModeMap[tlsMode]
}
```

Reply:
Noodling on how to alert the user a little more I have come to this position:

Most every product / application that gets deployed encourages users to check the logs for any errors or warning that occurred during the deploy. I say we ensure this is prominent in the documentation; that is is generally best practice to confirm the deployment by checking logs, monitors, etc. Do it.

Reply:
> I am not sure disregarding the whole configuration is a good idea...

I feel fairly strongly that partial application of configuration is not a good idea in cases like this. If the configuration is valid then it can be applied, if any part is invalid then ideally the application crashes and burns hard and maintains the previous configuration.  That way at any point, we know that the configuration that NLK is running is the last application that worked and there are not grey areas. 



> that it is generally best practice to confirm the deployment by checking logs, monitors, etc. Do it.

I agree, we should document usage this way.  However for the class of errors that we can catch at the configuration intake phase we should abort/error the application.  As we've talked about though I think starting with error logging will be a good start to a more robust strategy that catches both obvious misconfiguration as well as those that rear their heads post application.